File: code/python/nnesycef/Stamp.py
class Stamp(object):
    def __init__(self, arr):
        self._arr = arr

    def __getitem__(self, idx):
        return self._arr[idx]

    @staticmethod
    def merge(a, b):
        zipped = []

        ia = 0
        ib = 0

        for i in range(0, min(len(a._arr), len(b._arr))):
            if (i % 2) == 0:
                zipped.append(a[ia])
                ia+=1
            else:
                zipped.append(b[ib])
                ib+=1

        ## append remaining part of either stamp
        zipped.extend(a._arr[ia:])
        zipped.extend(b._arr[ib:])

        ## limit length
        zipped = zipped[:100]

        return Stamp(zipped)

    @staticmethod
    def checkOverlap(a, b):
        set_ = set()
        for iA in a._arr:
            set_.add(iA)

        for iB in b._arr:
            if iB in set_:
                return True

        return False

File: code/fsharp/NNeSyCeF/Stamp.fs
module Stamp
  type Stamp = struct
    val arr : int64[]
    new(arr_)={arr=arr_}
  end

  let STAMP_LENGTH = 200

  let merge (a:Stamp) (b:Stamp) =
    let commonLength = min (Array.length a.arr) (Array.length b.arr)

    // we need the subarrays of the lowest common length
    let zipArrA = Array.sub a.arr 0 commonLength
    let zipArrB = Array.sub b.arr 0 commonLength
  
    // and the remaining arrays
    let remainArrA = Array.sub a.arr commonLength ((Array.length a.arr) - commonLength)
    let remainArrB = Array.sub b.arr commonLength ((Array.length b.arr) - commonLength)

    let zipped = Array.zip zipArrA zipArrB |> Array.map (fun(a, b) -> [|a; b|]) |> Array.concat

    let completeZipped = [|zipped;remainArrA;remainArrB|] |> Array.concat

    // limit length to maximum stamp size
    let result =  Array.sub completeZipped 0 (min STAMP_LENGTH (Array.length completeZipped))

    Stamp(result)

  let checkOverlap (a:Stamp) (b:Stamp) =
    let aAsSet = Set.ofArray a.arr
    Array.exists (fun x -> Set.contains x aAsSet) b.arr
